docs/rfd/001.md

```diff
@@ -0,0 +1,181 @@
+
+The URI specification [RFC 3986](go/rfc/3986#section-3.3)[^1] defines a path segment as
+
+```
+segment = *pchar
+pchar = unreserved / pct-encoded / sub-delims / ":" / "@"
+unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
+sub-delims  = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
+pct-encoded = "%" HEXDIG HEXDIG
+```
+
+where HEXDIG should refer to a digit or an uppercase letter A-F. This matches the definition in https://url.spec.whatwg.org/#url-path-segment-string, which supposedly supersedes the RFC.
+
+Thus, define `EscapePathSegment` as a function that escapes all characters that are not unreserved, sub delimiters, `:` or `@`. See the Appendix for how this function can be implemented in some of the major programming languages.
+
+
+### Message Encoding
+
+Constructing a message takes three parameters: hostid, action and (optionally) hostid-type. The content to be embedded in the URL consists of 2 path segments: host-segment and action-segment.
+
+```
+host-segment = EscapePathSegment( [hostid-type) ':'] hostid )
+action-segment = EscapePathSegment(action)
+message = host-segment '/' action-segment
+tag = glome_tag(priv, pub, message)
+```
+
+The `hostid-type` prefix is added if and only if the `hostid-type` of the message is not empty.
+
+Note that this voids some of the existing recommendations for 'good' actions: `shell/root`, for example, would have to be escaped and thus be less readable. Instead, using URI sub-delimiters as in `shell=root` should be recommended. This format would interact nicely with a host-identity-based authorization scheme working with key-value pairs.
+
+
+### Message Decoding
+
+Take the URL prefix up to including the `/` after the handshake message. Split this string on the character `/` and keep only the first and second element, denoted `host-segment` and `action-segment`; or fail if there are less than two elements. Replace all percent-encoded octets in the `host-segment` with their raw, unencoded form. Split the result at the character `:`. If there is one element, assign that element to `hostid` and assign the literal string `hostname` to `hostid-type`; if there are two elements assign the first one to `hostid-type` and the second to `hostid`; otherwise fail. Replace all percent-encoded octets in the `action-segment` with their raw, unencoded form, and assign the result to `action`.
+
+
+## Exemplary Choice of Values
+
+*   Use 3 octets as server prefix (encodes cleanly into 4 base64 characters and is sufficiently safe not to cause problems even for keys that change frequently, see Appendix).
+*   Use only action 'shell'.
+
+With a short hostname `glome` and omitting the domain, the URL looks like http://glome/v2/lNXP/2TZm3tVBle2tHCbb6GLfMm5cpy1UcaMv7218zHKdCMw=/host1.int.example.com/shell/ and is 94 characters long. The corresponding GLOME Login v1 variant has 94 characters as well.
+
+
+# Alternatives considered
+
+## URL Format
+
+*   Encode both public keys in one base64 segment.
```

Comment:
I'd be in favour of keeping GLOME Login v1 behaviour of base64-encoding keys together. Separating them in a human-readable way (i.e., with `/`) brings no benefit to the user while increasing the URL size by two "fixed" characters (i.e., `/` and `=`). We might as well use 4 octets for the key prefix and further reduce the likelihood of a collision.